File: Calandar.java

import java.util.ArrayList;
import java.util.List;


class Calandar{
    private PresetManager presetManager;
    private Preset currentPreset;
    private String startDate;
    private String endDate;
    private ViewType viewType;
    /**
     * Constructor
     * @param presetManager list of group identifiers
     * @param startDate format "YYYY-MM-DDTHH:MM"
     * @param endDate   format "YYYY-MM-DDTHH:MM"
     * @param viewType  type of view {day, week, month}
     */
    public Calandar(PresetManager presetManager, String startDate, String endDate, ViewType viewType) {
        this.presetManager = presetManager;
        this.startDate = startDate;
        this.endDate = endDate;
        this.viewType = viewType;
    }

    public void setCurrentPreset(String presetName) {
        this.currentPreset = presetManager.getOrCreate(presetName);
    }

    public void addGroup(String presetName, Group group){
        Preset preset = presetManager.getOrCreate(presetName);
        preset.addGroup(group);
    }

    public void removeGroup(String presetName, Group group){
        Preset preset = presetManager.get(presetName);
        if (preset != null) {
            preset.removeGroup(group);
        }
    }

    /**
     * Recupere les informations de tous les cours dans la liste des groupes
     * @return list of cours
     * @throws Exception
     */
    public List<Cours> getCours() throws Exception{
        if (currentPreset == null) {
            throw new IllegalStateException("Pas de Preset selectionner");
        }
        CelcatToJson celcatToJson = new CelcatToJson("https://edt.uvsq.fr/Home/GetCalendarData", this.viewType);
        List<Cours> coursList = new ArrayList<>();
        for (Group group : currentPreset.getGroups()){
            String jsonData = celcatToJson.getCalendarData(this.startDate, this.endDate, "103", "M1 info", "3");
            JsonToCours jsonToCours = new JsonToCours(jsonData);
            coursList.addAll(
                jsonToCours.getCoursList().stream()
                    .filter(cours -> cours.getTitle().contains(group.getUECode()))//premier filtre sur UE
                    .filter(cours -> {
                        if (cours.isCM()) {
                            return true;
                        }
                        if (cours.isTD()) {
                            return cours.getGroup() != null
                                && cours.getGroup().equals(group.getGroupNumber());
                        }
                        return false;
                    })
                    .toList()//gère si c un CM osef du groupe mais si c un TD il le prend en compte
            );
        }
        return coursList;
    }
}

File: Discord_Bot/calandar.js
const celcatToJson = require('./celcatToJson');
const viewType = require('./viewType');
const jsonToCours = require('./jsonToCours');

class Calendar{
    /**
     * 
     * @param {[group]} groups 
     * @param {string} startDate 
     * @param {string} endDate 
     * @param {viewType} viewType 
     */
    constructor(groups, startDate, endDate, viewType){
        this.groups = groups;
        this.startDate = startDate;
        this.endDate = endDate;
        this.viewType = viewType;
    }

    addGroup(group){
        if (this.groups.indexOf(group) === -1){
            this.groups.push(group);
        }
    }
    removeGroup(group){
        const index = this.groups.indexOf(group);
        if (index !== -1){
            this.groups.splice(index, 1);
        }
    }

    getCoursWork(){
        return new Promise((resolve, reject) => {
            const coursList = [];
            for (let i = 0; i < this.groups.length; i++){
                const group = this.groups[i];
                const celcat = new celcatToJson("https://edt.uvsq.fr/Home/GetCalendarData", this.viewType);
                celcat.getCalandarData(this.startDate, this.endDate, '103', group.groupCode, '3').then(result => {
                    const cours = new jsonToCours(result).getCoursList();
                    for (const cour of cours){
                        if (cour.title.includes(group.UECode) && !coursList.includes(cour)){
                            coursList.push(cour);
                        }
                    }
                    // console.log(coursList);
                    resolve(coursList);
                }).catch(error => {
                    console.error("Error fetching calendar data:", error);
                    reject(error);
                });
            }
        });
    }
    async getCours(){
        const cours = await this.getCoursWork();
        return cours;
    }
}
module.exports = Calendar;